accepted/2021/User-Experience/Apple-Silicon-UX.md

```diff
@@ -0,0 +1,254 @@
+
+Overall the experiment demonstrated a highly workable user experience.
+
+- The experiment represents a viable design alternative. It did not support side-by-side install of multiple architectures of the same runtime version.
+
+- The implementation was incomplete. It did not automate the build process or deal with installer issues
+
+### Dotnet Publish Experience Inconsistent
+
+#### Signing
+
+- dotnet publish for Xamarin workloads handles signing automatically.
+- dotnet publish for CoreClr console apps do not handle signing automatically. The developer would need to explicitly sign after publishing for Apple Silicon.  The signing command is relatively simple `codesign -s <signature> --entitlements <app-entitlements-path> <app-path>`.
+- There is no message to the customer that the published app needs to be signed
+
+#### Dotnet Publish as Universal Binaries
+
+- This scenario is also supported for Xamarin iOS workloads. This is a slightly different scenario as Xamarin iOS apps are native binaries (not JIT binaries)
+- CoreCLR runtime apps have no mechanism to publish universal binaries.
+
+### Dotnet-SOS and the SOS plugin Issues
+
+The `dotnet-sos` tool installs the `SOS` diagnostic plugin into `lldb`. Since `lldb` is a universal binary, it expects its plugins to also be universal binaries. With the single architecture design, when `SOS` is installed, `lldb` will complain about not being able to load the plugin when the architecture mismatches.
+
+#### Universal Binary Experiment
+
+Making the Apple Silcon SOS be a universal binary to allowed SOS to work on `x86_64` and `arm64` processes.
+
+## User Experience Designs
+
+### Mixed Architecture Side By Side Installs
+
+This is controversial. This is _uncommitted/unplanned_. Ongoing discussions will likely affect this proposal.
+
+For CoreCLR runtime, the proposal would be to:
+
+- Make `dotnet` & `hostfxr` to be universal binaries
+- Support automatic architecture selection
+  - Modify `hostfxr` to assume .NET 5 and earlier require x64 architecture.
+  - Modify `dotnet` to restart in the correct architecture on an architecture mismatch.
+- Make the CoreCLR runtime effectively a Universal Binaries
+  - Create a single distribution for both macOS architectures.
+  - Make the runtime native binaries to be universal binaries
+  - Modify `hostpolicy` to probe for trusted platform assemblies based on the current RID. This is designed to architecture specific ready-to-run managed assemblies. This RID specific path would be probed before probing the common any platform assembly path.
+  - Modify the runtime probing to be consistent with the new `hostpolicy` probing.
+- Modify the build system to build the Universal Binaries
+- Modify installers
+  - Modify .NET 5 installers to allow installation without breaking universal `dotnet`.
```

Comment:
```suggestion
  - Modify .NET 3.1 and .NET 5 installers to allow installation without breaking universal `dotnet`.
```
? .NET 5 goes out of support short after .NET 6 ships. .NET 3.1 is supported for much longer.